R/intervals.R

```diff
@@ -0,0 +1,625 @@
+#'   intervals in `ints_dt`.
+#' @param full_int_end \[`numeric(1)`\]\cr
+#'   End of the complete interval that should be covered by the detailed
+#'   intervals in `ints_dt`.
+#'
+#' @return  \[`data.table()`\] with columns for 'start' and 'end' of intervals
+#'   where intervals in `ints_dt` are missing or overlapping. If no intervals
+#'   are missing or overlapping then zero-row \[`data.table()`\] is returned.
+#'
+#' @rdname problematic_intervals
+identify_missing_intervals <- function(ints_dt,
+                                       full_int_start = 0,
+                                       full_int_end = Inf) {
+
+  assertive::is_data.table(ints_dt)
+  assertthat::is.number(full_int_start)
+  assertthat::is.number(full_int_end)
+
+  # create full interval that all sub intervals should span
+  full_int <- intervals::Intervals_full(
+    matrix(c(full_int_start, full_int_end), ncol = 2), closed = c(TRUE, FALSE)
+  )
+
+  # create left-closed, right-open intervals
+  ints <- intervals::Intervals_full(as.matrix(ints_dt), closed = c(TRUE, FALSE))
+
+  # identify missing intervals
+  missing_ints <- intervals::interval_difference(full_int, ints)
+
+  missing_ints_dt <- data.table::as.data.table(missing_ints)
+  data.table::setnames(missing_ints_dt, c("start", "end"))
+  data.table::setkeyv(missing_ints_dt, c("start", "end"))
+  return(missing_ints_dt)
+}
+
+#' @rdname problematic_intervals
+identify_overlapping_intervals <- function(ints_dt,
+                                           full_int_start = 0,
+                                           full_int_end = Inf) {
+
+  assertive::is_data.table(ints_dt)
+  assertthat::is.number(full_int_start)
+  assertthat::is.number(full_int_end)
+
+  # create left-closed, right-open intervals
+  ints <- intervals::Intervals_full(as.matrix(ints_dt), closed = c(TRUE, FALSE))
+
+  # get list mapping between intervals if they overlap at all
+  overlaps <- intervals::interval_overlap(ints, ints)
+  names(overlaps) <- 1:length(overlaps)
+
+  # sort by number of intervals that each interval overlaps with so that we can
+  # identify the largest overlapping intervals first
+  overlaps <- overlaps[order(sapply(overlaps, length), decreasing=T)]
+
+  overlapping_indices <- c()
+  for (i in names(overlaps)) {
+    # remove match to itself
+    overlaps[[i]] <- overlaps[[i]][overlaps[[i]] != i]
+
+    # remove indices of overlapping intervals that have already been identified
+    overlaps[[i]] <- overlaps[[i]][!overlaps[[i]] %in% overlapping_indices]
+
+    if (length(overlaps[[i]]) > 0) {
+      overlapping_indices <- c(overlapping_indices, i)
+    }
+  }
+  overlapping_ints <- ints[as.integer(overlapping_indices)]
+
+  overlapping_ints_dt <- data.table::as.data.table(overlapping_ints)
+  data.table::setnames(overlapping_ints_dt, c("start", "end"))
+  data.table::setkeyv(overlapping_ints_dt, c("start", "end"))
+  return(overlapping_ints_dt)
+}
+
+#' @title create new interval nodes for any missing interval nodess
```

Comment:
typo "nodes"

Reply:
Also, we should make sure terms like "node" are clearly defined in the package documentation.

Reply:
I explained it briefly in the vignette. I think users would only need to know about the term if debugging. 

I probably need to do a sweep of the exported functions and make sure they aren't mentioned much if at all

Reply:
I added some links to wikipedia explaining what tree data structures are too